stack-graphs/src/cycles.rs

```diff
@@ -127,3 +135,150 @@ where
         true
     }
 }
+
+// ----------------------------------------------------------------------------
+// Cycle detector
+
+pub trait Appendable {
+    type Ctx;
+
+    fn append_to(
+        &self,
+        graph: &StackGraph,
+        partials: &mut PartialPaths,
+        ctx: &mut Self::Ctx,
+        path: &mut PartialPath,
+    ) -> Result<(), PathResolutionError>;
+    fn start_node(&self, ctx: &mut Self::Ctx) -> Handle<Node>;
+    fn end_node(&self, ctx: &mut Self::Ctx) -> Handle<Node>;
+}
+
+#[derive(Clone)]
+pub struct AppendingCycleDetector<A> {
+    appendages: List<A>,
+}
+
+pub type Appendages<A> = ListArena<A>;
+
+impl<A: Appendable + Clone> AppendingCycleDetector<A> {
+    pub fn new() -> Self {
+        Self {
+            appendages: List::empty(),
+        }
+    }
+
+    pub fn from(appendages: &mut Appendages<A>, appendage: A) -> Self {
```

Comment:
If you follow the above suggestion then the final parameter would become `appendable: A` as well.